akka-actor/src/main/scala/akka/actor/ActorSystem.scala

```diff
@@ -1292,4 +1304,28 @@ private[akka] class ActorSystemImpl(
      */
     def terminationFuture: Future[T] = done.future
   }
+
+  // like a lazy val but without synchronization
+  // 1             -> typed guardian
+  // 2             -> not a typed guardian
+  // anything else -> undetermined, use mild reflection to check
+  //
+  // It's OK to check multiple times (result will be the same), but we want to eventually
+  // stop checking
+  private var _isTypedGuardian: Int = 0
+
+  private def isTypedGuardian: Boolean =
+    _isTypedGuardian match {
+      case 1 => true
+      case 2 => false
+      case _ =>
+        if (guardianProps.exists(_.clazz.getName.startsWith("akka.actor.typed"))) {
```

Comment:
`isTypedGuardian` isn't used in any hot-path as far as I can see so no need for optimizations (that are more difficult to read). Could be a lazy val but I don't see that any of this is much expensive so could just be a `def` without any caching?

Reply:
Yeah, I guess in the failure case, being slow and thus introducing some backpressure is a good thing.